### PR TITLE
Do some git gc fixes for Keno?

### PR DESCRIPTION
**Pull request by kshyatt**
https://github.com/JuliaLang/julia/issues/24464 trying to address some of this, hopefully I did it semi-correctly...

### CODE REVIEW

File: base/libgit2/commit.jl [hunk 1]
```diff
@@ -15,7 +15,8 @@ function message(c::GitCommit, raw::Bool=false)
     if msg_ptr == C_NULL
         return nothing
     end
-    return unsafe_string(msg_ptr)
+    msg_str = Base.@gc_preserve c unsafe_string(msg_ptr)
```
Comment (Keno) [inline base/libgit2/commit.jl]: This is fine for the compiler as is, but probably best to wrap the entire function (from before the ccall to after the unsafe_string) in `gc_preserve c` to make sure we get the scope correct.

File: base/libgit2/commit.jl [hunk 2]
```diff
@@ -27,7 +28,8 @@ the person who made changes to the relevant file(s). See also [`committer`](@ref
 function author(c::GitCommit)
     ptr = ccall((:git_commit_author, :libgit2), Ptr{SignatureStruct}, (Ptr{Void},), c.ptr)
     @assert ptr != C_NULL
-    return Signature(ptr)
+    sig = Base.@gc_preserve c Signature(ptr)
```
Comment (Keno) [inline base/libgit2/commit.jl]: same

File: base/libgit2/commit.jl [hunk 3]
```diff
@@ -40,8 +42,8 @@ a `committer` who committed it.
 """
 function committer(c::GitCommit)
     ptr = ccall((:git_commit_committer, :libgit2), Ptr{SignatureStruct}, (Ptr{Void},), c.ptr)
-    @assert ptr != C_NULL
-    return Signature(ptr)
+    sig = Base.@gc_preserve c Signature(ptr)
```
Comment (Keno) [inline base/libgit2/commit.jl]: same

File: base/libgit2/commit.jl (hunk 1 above)
Comment (kshyatt) [reply base/libgit2/commit.jl]: OK, I'll go in and fix these. Thanks for the assist!

**Comment by kshyatt** [comment]
Better?

File: base/libgit2/commit.jl [hunk 4]
```diff
@@ -9,13 +9,16 @@ leading newlines removed). If `raw` is `true`, the message is not stripped
 of any such newlines.
 """
 function message(c::GitCommit, raw::Bool=false)
+    c_flag = Base.@_gc_preserve_begin c
```
Comment (Keno) [inline base/libgit2/commit.jl]: Can be written as 
```
@gc_preserve c begin

end
```
The underscore versions are only for early in the code base where that macro does not exist yet.

**Review by Keno — APPROVED** [review]

**Pull request merged**